src/v/cloud_topics/level_one/frontend_reader/reader.cc

```diff
@@ -0,0 +1,415 @@
+          "Failed to open stream for L1 object {} for {} ({}): {}",
+          _current_obj->oid,
+          _ntp,
+          _tidp,
+          std::to_underlying(stream_result.error()));
+        _state = state::end_of_stream;
+        co_return;
+    }
+
+    auto reader = l1::object_reader::create(std::move(stream_result).value());
+    while (true) {
+        auto result_fut = co_await ss::coroutine::as_future(
+          reader->read_next());
+        if (result_fut.failed()) {
+            vlog(
+              cd_log.error,
+              "Exception reading L1 object {} for {} ({}): {}",
+              _current_obj->oid,
+              _ntp,
+              _tidp,
+              result_fut.get_exception());
+            co_await reader->close();
+            _state = state::end_of_stream;
+            co_return;
+        }
+
+        auto result = result_fut.get();
+        if (std::holds_alternative<model::record_batch>(result)) {
+            auto batch = std::move(std::get<model::record_batch>(result));
+
+            // Skip batches before our start offset.
+            if (
+              batch.last_offset() < kafka::offset_cast(_config.start_offset)) {
+                continue;
+            }
+
+            // Stop if we've reached our max offset.
+            if (batch.base_offset() >= kafka::offset_cast(_config.max_offset)) {
+                break;
+            }
+
+            // Check if adding this batch would exceed our byte limit.
+            size_t batch_size = batch.size_bytes();
+            if (
+              (_config.strict_max_bytes || _config.bytes_consumed > 0)
+              && _config.bytes_consumed + batch_size >= _config.max_bytes) {
+                _config.over_budget = true;
+                break;
+            }
+            _config.bytes_consumed += batch_size;
+
+            // If we make it past all that, emit the batch.
+            _batches.push_back(std::move(batch));
+        } else if (std::holds_alternative<model::topic_id_partition>(result)) {
+            // Partition marker. Done with this ntp's partition.
+            break;
+        } else if (
+          std::holds_alternative<l1::footer>(result)
+          || std::holds_alternative<l1::object_reader::eof>(result)) {
+            // End of data.
+            break;
+        }
+    }
+
+    co_await reader->close();
+
+    // Note that it's possible to materialize zero batches.
+    vlog(
+      cd_log.debug,
+      "Materialized {} batches from L1 object {} for {} ({})",
+      _batches.size(),
+      _current_obj->oid,
+      _ntp,
+      _tidp);
+    _state = state::materialized;
+}
+
+void level_one_log_reader_impl::consume_materialized_batches(
+  chunked_circular_buffer<model::record_batch>* dest) {
+    vlog(
+      cd_log.debug,
+      "Consuming {} materialized batches for {} ({})",
+      _batches.size(),
+      _ntp,
+      _tidp);
+
+    std::move(_batches.begin(), _batches.end(), std::back_inserter(*dest));
+    _batches.clear();
```

Comment:
why not keep _batches in the same data structure as `dest` so they can just be swapped out instead of just moving the container?